Pen children

adamnovak opened the pull request:
Fixed #3045.

## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 * Single machine batch system keeps better track of child processes so their children can be stopped

## Reviewer Checklist

<!-- To be kept in sync with docs/contributing/checklist.rst -->

 * [x] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [ ] Read through the code changes. Make sure that it doesn't have:
    * [ ] Addition of trailing whitespace.
    * [ ] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [ ] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [ ] New functions or classes without informative docstrings.
    * [ ] Changes to semantics not reflected in the relevant docstrings.
    * [ ] New or changed command line options for Toil workflows that are not reflected in `docs/running/cliOptions.rst`
    * [ ] New features without tests.
* [ ] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [ ] Finish the review with an overall description of your opinion.

## Merger Checklist

<!-- To be kept in sync with docs/contributing/checklist.rst -->

* [x] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

mr-c commented on src/toil/test/batchSystems/batchSystemTest.py:
```diff
@@ -415,6 +418,29 @@ def testIgnoreNode(self):
+    Dump a string into a temp file and return its path.
+    """
+    fd, path = tempfile.mkstemp()
+    try:
```
Please use pytest for temp file creation, so that they get deleted after the tests

adamnovak replied:
I can't use `pytest`'s fixtures to manage temporary files here; they aren't available in `unittest.TestCase` tests like all the Toil tests are.

But, Toil already has `_createTempDir()` and `_getScriptSource()` methods on its test case base class, which is going to be the Right Way to do those things, so I am using those now.

mr-c commented:
>     * Single machine batch system tries to corral child processes in groups

Cool, but confusing for a user reading the changelog🙂. Maybe state the reason for the change? (Less/no processes remaining after termination)

adamnovak commented:
OK @mr-c, I have started using Toil's centralized way to get temp directories for tests, and improved the changelog entry.

mr-c commented on src/toil/test/batchSystems/batchSystemTest.py:
```diff
@@ -415,6 +418,22 @@ def testIgnoreNode(self):
+    """
+    Dump a string into a temp file and return its path.
+    """
+    fd, path = tempfile.mkstemp(dir=temp_dir)
```
Shouldn't `_createTempDir` be used here?

mr-c commented on src/toil/batchSystems/singleMachine.py:
```diff
@@ -261,6 +259,48 @@ def _checkOnDaddy(self):
             self.daddyException = None
             raise exc
 
+    def _stop_now(self, popens: List[subprocess.Popen]):
```
Missing a return value on the type hints

mr-c commented on src/toil/batchSystems/singleMachine.py:
```diff
@@ -261,6 +259,48 @@ def _checkOnDaddy(self):
+                # never managed to make the group.
+                popen.kill()
+
+    def _stop_and_wait(self, popens: List[subprocess.Popen]):
```
Return value also missing here

mr-c reviewed (changes requested):
Here is the result of `make diff-mypy` from https://github.com/DataBiosphere/toil/pull/3492
```
src/toil/batchSystems/singleMachine.py (48.3%): Missing lines 242,244,262,285,292,298,443-444,517,521,588,596,598-599,601
src/toil/jobStores/aws/jobStore.py (23.1%): Missing lines 1202,1206-1211,1213-1214,1220
src/toil/test/batchSystems/batchSystemTest.py (52.5%): Missing lines 450,459,467-468,470,475,482,487,489,493-495,497,499,501,504,506,509,518,520,535,541,545,547,552,558,576,580,619
-------------
Total:   103 lines
Missing: 54 lines
Coverage: 47%
-------------
```

adamnovak commented:
I've added annotations for the lack of a return value.

Some of the mypy complaints are about lines like this one, where it's not clear that any type annotation would make sense to have:

https://github.com/DataBiosphere/toil/blob/082d118f36bc2b586aba3b731758d68301c9c8f0/src/toil/batchSystems/singleMachine.py#L242

I've addressed some places where it looks like adding a type annotation on a variable might help, and changed some annotations that were more restrictive than they should have been. But every annotation you add is going to spawn off a mypy complaint about every line where you fill in that value from something not annotated to be the correct type, right?

I don't think we are or should be trying to ensure that we add no new mypy type checking complaints, just no new functions without type hints on their signatures. If we want to expand that, we could say that we don't want to add methods or fields without type annotations.

It wouldn't be appropriate for me to go much further in adding new type hints as part of this PR; chunks of the code that don't have type hints but should would want to be their own tickets.

adamnovak commented:
Looks like we managed to dig up a failure in the file job store's job update logic:

```
[2021-03-25T00:52:07+0000] [MainThread] [I] [toil.worker] Redirecting logging to /tmp/node-ce3ecfc4-0724-41c2-bba3-9dd4c1392652-5134a9c7ae97b9c469839b82601897f4/tmpmw_e_2d6/worker_log.txt
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/builds/databiosphere/toil/src/toil/fileStores/cachingFileStore.py", line 1777, in startCommitThread
    self.jobStore.update(self.jobDesc)
  File "/builds/databiosphere/toil/src/toil/jobStores/fileJobStore.py", line 235, in update
    os.rename(self._getJobFileName(job.jobStoreID) + ".new", self._getJobFileName(job.jobStoreID))
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/toil-test-toil.test.src.jobFileStoreTest.JobFileStoreTest-testJobFileStore-jobstore-dqyj15re/jobs/kind-JobFunctionWrappingJob/instance-l1k_mkmn/job.new' -> '/tmp/toil-test-toil.test.src.jobFileStoreTest.JobFileStoreTest-testJobFileStore-jobstore-dqyj15re/jobs/kind-JobFunctionWrappingJob/instance-l1k_mkmn/job'
[2021-03-25T00:52:09+0000] [Thread-18 ] [D] [toil.fileStores.cachingFileStore] Committing file uploads asynchronously
```

Also our stack trace is missing a line, and in a weird place.
Here's what the leader was doing:

```
DEBUG    toil.serviceManager:serviceManager.py:31 Initializing service manager
DEBUG    toil.leader:leader.py:507 Built the jobs list, currently have 1 jobs to update and 0 jobs issued
DEBUG    toil.leader:leader.py:433 Updating status of job 'JobFunctionWrappingJob' kind-JobFunctionWrappingJob/instance-l1k_mkmn with result status: 0
DEBUG    toil.batchSystems.singleMachine:singleMachine.py:546 Issuing the command: _toil_worker JobFunctionWrappingJob file:/tmp/toil-test-toil.test.src.jobFileStoreTest.JobFileStoreTest-testJobFileStore-jobstore-dqyj15re kind-JobFunctionWrappingJob/instance-l1k_mkmn with memory: 2147483648, cores: 1.0, disk: 2147483648
INFO     toil.leader:leader.py:740 Issued job 'JobFunctionWrappingJob' kind-JobFunctionWrappingJob/instance-l1k_mkmn with job batch system ID: 0 and cores: 1, disk: 2.0 G, and memory: 2.0 G
DEBUG    toil.batchSystems.singleMachine:singleMachine.py:464 Launched job 0 as child 14158
INFO     toil.leader:leader.py:819 1 jobs are running, 0 jobs are issued and waiting to run
ERROR    toil.batchSystems.singleMachine:singleMachine.py:518 Got exit code 1 (indicating failure) from job _toil_worker JobFunctionWrappingJob file:/tmp/toil-test-toil.test.src.jobFileStoreTest.JobFileStoreTest-testJobFileStore-jobstore-dqyj15re kind-JobFunctionWrappingJob/instance-l1k_mkmn.
DEBUG    toil.batchSystems.singleMachine:singleMachine.py:530 Child 14158 for job 0 succeeded
DEBUG    toil.batchSystems.singleMachine:singleMachine.py:632 Ran jobID: 0 with exit value: 1
WARNING  toil.leader:leader.py:570 Job failed with exit value 1: 'JobFunctionWrappingJob' kind-JobFunctionWrappingJob/instance-l1k_mkmn
Exit reason: None
DEBUG    toil.leader:leader.py:957 Job 'JobFunctionWrappingJob' kind-JobFunctionWrappingJob/instance-l1k_mkmn continues to exist (i.e. has more to do)
WARNING  toil.leader:leader.py:989 No log file is present, despite job failing: 'JobFunctionWrappingJob' kind-JobFunctionWrappingJob/instance-l1k_mkmn
WARNING  toil.job:job.py:742 Due to failure we are reducing the remaining try count of job 'JobFunctionWrappingJob' kind-JobFunctionWrappingJob/instance-l1k_mkmn with ID kind-JobFunctionWrappingJob/instance-l1k_mkmn to 0
```

I have no idea why it is doing this. It doesn't seem really related to this PR, but I've also never seen it before.

adamnovak commented:
@mr-c I think I got all the missing return values, and this is the next PR in the queue to merge, going old to new. Can you sign off on this?

DailyDreaming reviewed (approved):
This looks great.  Thanks for adding all of the typing.

adamnovak commented:
It looks like whenever this was last being tested AWS just wasn't handing out instances promptly, causing the tests for the (unmodified) provisioner to fail. I can't really replicate the issue now.

adamnovak commented:
Since this is due today, and the changes on master seem orthogonal, I am going to merge this if the re-run provisioner tests pass: https://ucsc-ci.com/databiosphere/toil/-/jobs/81986

adamnovak commented:
OK, it succeeded, but I found a worrying looking test failure in one of the other runs, so I've now hardened the code against cases where SIGKILLs might still be in flight, at the cost of greater vulnerability to PID/PGID re-use issues.

There really isn't a correct API for process management that lets you guarantee that a process group is gone and hasn't been replaced, as far as I can tell.
